packages/runtime/test/integration/atomic-transfer.ts

```diff
@@ -0,0 +1,431 @@
+
+    syncAccounts();
+    assert.equal(john.balance(), initialBalance - 2100);
+    assert.equal(alice.balance(), initialBalance + 100);
+  });
+
+  it("should fail if payment transaction fails", () => {
+    const txGroup: ExecParams[] = [
+      {
+        type: TransactionType.CallNoOpSSC,
+        sign: SignType.SecretKey,
+        fromAccount: john.account,
+        appId: appId,
+        payFlags: { totalFee: 1000 }
+      },
+      {
+        type: TransactionType.TransferAlgo,
+        sign: SignType.SecretKey,
+        fromAccount: alice.account,
+        toAccountAddr: john.address,
+        amountMicroAlgos: 6e6,
+        payFlags: { totalFee: 1000 }
+      }
+    ];
+    expectRuntimeError(
+      () => runtime.executeTx(txGroup),
+      RUNTIME_ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_BALANCE
+    );
+
+    const localCounter = runtime.getAccount(john.address).getLocalState(appId, key);
+    assert.isDefined(localCounter);
+    assert.equal(localCounter, 0n);
+
+    syncAccounts();
+    assert.equal(john.balance(), initialBalance);
+    assert.equal(alice.balance(), initialBalance);
+  });
+
+  it("should not freeze asset if payment fails", () => {
+    const txGroup: ExecParams[] = [
+      {
+        type: TransactionType.FreezeAsset,
+        sign: SignType.SecretKey,
+        fromAccount: john.account,
+        assetID: assetId,
+        freezeTarget: alice.address,
+        freezeState: true,
+        payFlags: {}
+      },
+      {
+        type: TransactionType.TransferAlgo,
+        sign: SignType.SecretKey,
+        fromAccount: alice.account,
+        toAccountAddr: john.address,
+        amountMicroAlgos: 6e6,
+        payFlags: { totalFee: 1000 }
+      }
+    ];
+
+    expectRuntimeError(
+      () => runtime.executeTx(txGroup),
+      RUNTIME_ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_BALANCE
+    );
+
+    syncAccounts();
+    assert.equal(alice.getAssetHolding(assetId)?.["is-frozen"], false);
+  });
+
+  it("should not modify asset if payment fails", () => {
+    const modFields = {
+      manager: john.address,
+      reserve: john.address,
+      clawback: john.address,
+      freeze: john.address
+    };
+    const txGroup: ExecParams[] = [
+      {
+        type: TransactionType.ModifyAsset,
+        sign: SignType.SecretKey,
+        fromAccount: john.account,
+        assetID: assetId,
+        fields: modFields,
+        payFlags: {}
+      },
+      {
+        type: TransactionType.TransferAlgo,
+        sign: SignType.SecretKey,
+        fromAccount: alice.account,
+        toAccountAddr: john.address,
+        amountMicroAlgos: 6e6,
+        payFlags: { totalFee: 1000 }
+      }
+    ];
+
+    expectRuntimeError(
+      () => runtime.executeTx(txGroup),
+      RUNTIME_ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_BALANCE
+    );
+
+    // Verify asset manager is not changed
+    assert.equal(runtime.getAssetDef(assetId).manager, "WHVQXVVCQAD7WX3HHFKNVUL3MOANX3BYXXMEEJEJWOZNRXJNTN7LTNPSTY");
```

Comment:
Can we define `assetManagerOrig = runtime.getAssetDef(assetId).manager` before `executeTx()` and use here ?

Reply:
yes